lib/uplink/main.go

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2018 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package uplink
+
+import (
+	"context"
+
+	"storj.io/storj/pkg/transport"
+
+	"storj.io/storj/pkg/identity"
+	"storj.io/storj/pkg/metainfo/kvmetainfo"
+	"storj.io/storj/pkg/miniogw"
+	"storj.io/storj/pkg/peertls/tlsopts"
+	"storj.io/storj/pkg/storj"
+	ul "storj.io/storj/uplink"
+)
+
+// Config holds the configs for the Uplink
+type Config struct {
+	// MaxBufferMem controls upload performance and is system-specific
+	MaxBufferMem int
+
+	// These should only be relevant for new files; these values for existing
+	// files should come from the metainfo index. It's unlikely these will ever
+	// change much.
+	EncBlockSize  int
+	MaxInlineSize int
+	SegmentSize   int64
+	TLSConfig     tlsopts.Config
+}
+
+// Session represents a specific access session.
+type Session struct {
+	TransportClient transport.Client
+	Gateway         *miniogw.Gateway
+}
+
+// Access is all of the access information an application needs to store and
+// retrieve data. Someone with a share may have no restrictions within a project
+// (can create buckets, list buckets, list files, upload files, delete files,
+// etc), may be restricted to a single bucket, may be restricted to a prefix
+// within a bucket, or may even be restricted to a single file within a bucket.
+// NB(dylan): You need an Access to start a Session
+type Access struct {
+	Permissions Macaroon
+
+	// TODO: these should be per-bucket somehow maybe? oh man what a nightmare
+	// Could be done via []Bucket with struct that has each of these
+	// PathCipher       storj.Cipher // i.e. storj.AESGCM
+	// EncPathPrefix    storj.Path
+	// Key              storj.Key
+	// EncryptionScheme storj.EncryptionScheme
+
+	// Something like this?
+	// TODO(dylan): Shouldn't actually use string, this is just a placeholder
+	// until a more precise type is figured out - probably type Bucket
+	Buckets map[string]BucketOpts
+}
+
+// Uplink represents the main entrypoint to Storj V3. An Uplink connects to
+// a specific Satellite and caches connections and resources, allowing one to
+// create sessions delineated by specific access controls.
+type Uplink struct {
+	ID            *identity.FullIdentity
+	Session       *Session
+	SatelliteAddr string
+	Config        Config
+}
+
+// NewUplink creates a new Uplink
+func NewUplink(ident *identity.FullIdentity, satelliteAddr string, cfg Config) *Uplink {
+	return &Uplink{
+		ID:            ident,
+		SatelliteAddr: satelliteAddr,
+		Config:        cfg,
+	}
+}
+
+// ParseAccess parses a serialized Access
+func ParseAccess(data []byte) (Access, error) {
+	panic("TODO")
+}
+
+// Serialize serializes an Access message
+func (a *Access) Serialize() ([]byte, error) {
+	panic("TODO")
```

Comment:
Is serializing an Access message a thing we want to do? Would the Access struct still include the encryption key and everything? I'm kind of thinking that the only extra thing that you might need to serialize (beyond the normal GRPC stuff) is the macaroons.

Reply:
This is a hold over from the libuplink-draft branch PR that I was shown for this ticket. I'm not actually sure if we want to do this or not yet.

Reply:
yeah, we need encryption details if you are only shared a subfolder

Reply:
Oh, I get it, this is for communicating details of a share with another user, not for directly participating in the network. Cool then